### ISSUE TITLE
[Feature Request] Auto open databases like KeeAutoExec

### ISSUE THREAD

**Issue by bokysan** [keeweb/keeweb#321]
Hi,

since KeeWeb supports multiple databases by design, it would be great if it could open them up automatically in the same manner as does KeeAutoExec: http://keepass.info/plugins.html#keeautoexec.

I'm split between this being a plugin or behaviour by default, but given that KeeWeb also runs as a webapp, having this feature embedded does make some kind of sense.

From plugin's manual:
> Usage
>
> First of all, create a new group in your database and name it "AutoOpen" (without the quotes).
> 
> Each non-expired entry in this group corresponds to a database that should be opened automatically when the current database is opened. The fields of each entry specify the following:
> 
> *Title:* Ignored by the plugin; can be used for comments.
> *User name:* Must contain the path to the key file, if one should be used. The path can be either absolute or relative to the directory containing 'KeePass.exe'.
> *Password:* The master password for the database to open. If no password is required, leave this field empty.
> *URL:* Path to the database to open. The path can be either absolute or relative to the directory containing 'KeePass.exe'.
> *Notes:* Ignored by the plugin; can be used for comments.
> *Credentials for URLs.* When opening a database from an URL, the credentials for accessing the file on the server can be specified using the custom entry strings "IocUserName" and "IocPassword" (without the quotes).

> *Existence check.* If you create a custom entry string field named "Skip if not exists" and value "True" (without the quotes), the plugin will not attempt to open the database if it doesn't exist (no warning message will be shown in this case).

> *Direct key file.* If the entry has a file attachment named "KeyFile.bin" (without the quotes), this is used as key file (only if no key file path has already been specified using the user name field).

> *Order.* If multiple databases should be opened, KeeAutoExec by default opens them in the order that the entries are stored in the database (in 'unsorted' mode; you need to turn off sorting by a column to see this order). Anyway, you also can explicitly specify/override the order by defining priorities. To do this, create a custom entry string field "Priority" (without the quotes) in each entry and set it to a number. Lower numbers mean higher priority; higher numbers mean low priority. Negative numbers are allowed. Example: you define an entry for database A.kdbx with the "Priority" field set to 50, a B.kdbx entry with priority 100, and a C.kdbx entry with priority -10. Then, KeeAutoExec would open the databases in order C.kdbx, A.kdbx, B.kdbx. If you don't specify a priority, KeePass assigns consecutive numbers, starting from 0 (i.e. the first database has a priority of 0, the second one 1, etc.).

> *Placeholders.* All strings are Spr-compiled, i.e. placeholders and environment variables can be used.

**Comment by antelle** [comment]
This makes sense :+1: 
This way, one file can serve a master file, by opening which you will automatically get access to linked files. But sure, we must implement it not as adding group manually, this should be done under the hood and be configurable in file properties, e.g. 'add linked file' button.

**Comment by bokysan** [comment]
Well, if you start off by just supporting the feature on existing files and plastering the GUI on top of it later on, I'd be more than happy. :-D

**Comment by antelle** [comment]
No, we'll not implement features without UI. It's better to not have the feature at all.

**Comment by bokysan** [comment]
Ok!

Thanks.

B

**Comment by scic** [comment]
:thumbsup: I would find this a very useful feature as well.

**Comment by sebastianroming** [comment]
Any updates on this feature? This is the only reason I can't use KeeWeb at the moment... would love to use it.

**Comment by antelle** [comment]
No updated for now, but now it's easier to implement, since we support plugins. Later this could be included in product.

**Comment by pwFoo** [comment]
I would need that feature to. Important for desktop app and a nice to have for the web app.

**Comment by haraldreingruber** [comment]
👍 for this feature. It would be really useful for sharing database files within different teams (Users vs Admins).

**Comment by gilrim** [comment]
+1 for this

**Comment by gilrim** [comment]
Hi, just wanted to extend on the reasoning behind AutoOpen.

We use it to allow for both shared and personal passwords in keepass. By having a private.kdbx file that have the key to unlock a group file, and assigning new passwords to that file; we've got some sort of access control at the same time. This is of great value to my team and would allow us to switch to keeweb.

**Comment by thecount2a** [comment]
This seems to be a highly requested feature.  I would also really like this. Can this be implemented as a plugin? I assume it can since plugins seem to allow arbitrary JavaScript to be injected although we still have to consider whether all hooks that are needed are currently available for plugins. Can any devs comment on how this ought to be approached? I don't know the keepass codebase (yet) but I know JavaScript and maybe I could implement this.

**Comment by Coises** [comment]
Would it make sense to split this into two parts?

1. Have a kind of entry in KeeWeb that represents a KeeWeb data base, with the type (local file, WebDAV, Google Drive, etc.), path and credentials needed to access the file, and the password.

2. Have an option, either as part of the entry or as a separate list somewhere else in the user interface, to open such databases automatically.

Here is the use case for making it possible to have 1 without 2:

My partner and I each have KeeWeb databases. Since we both have entries for many of the same things (like Amazon), it would be annoying to have all entries in one database, or to have both databases open all the time, as autotype would always have to stop to ask us which entry we wanted.

Yet on occasion one of us might need to get to information from the other’s database. I’ve put an entry in her database that gives the information for how to get to my database. Having accessed it once, it’s on the quick list on the open database screen, so normally she only has to put in the password.

Even for that case, though, she has to copy the password to the clipboard (insecure), then go to the open screen, pick the right database and paste the password. If, for some reason, she should lose the entry in the quick list, it would be quite an annoyance to have to copy all the information to get to the WebDAV file and enter that.

It would be much nicer if she could just click a link in the entry for my database which would open it using the saved storage type, path, credentials and password.

The tricky bit, into which I have no insight, would be how to keep this consistent and inter-operable with KeePass. One thing we certainly do not want to do is create any incompatibilities with their database format.

**Comment by bokysan** [comment]
> Have a kind of entry in KeeWeb that represents a KeeWeb data base, with the type (local file, WebDAV, Google Drive, etc.), path and credentials needed to access the file, and the password.

It would definitely be sensible to keep the compatibility with KeeAutoExec in this manner. But please read up on the [comment above](#issuecomment-240997140) -- devs are against implementing the features without the GUI, and I can understand them about it.

**Comment by Coises** [comment]
In my comment about consistency I wasn’t thinking of compatibility with KeeAutoExec so much as I was wondering if the KeePass database format already includes a standard way of defining an entry whose purpose is to open (not *auto*-open) another database.

If there is, and if it’s flexible enough to encompass the storage types with which KeeWeb works, that would really be the place to start.

If there isn’t, then even if we avoid breaking other programs that can read KeePass databases, we risk creating a “private language” for KeeWeb. Essentially, that’s what it sounds like KeeAutoExec has done. I presume databases built with the special KeeAutoExec entries can be read by any program that reads KeePass databases; but the entries won’t have the meaning that they do when the extension is in use.

That might be an acceptable downside, but it is a consideration to weigh, especially if the feature would be part of the program and not an extension.

My main point was just that it would be nice to be able to open — as opposed to auto-open — other databases, and that it might make sense to use opening as a stepping stone to auto-opening.

Now that I look up the full documentation on KeeAutoExec, I see that it does have a way of setting entries that don’t auto-open (Enabled: False) and that it creates a File menu command to open databases manually. That covers my use case, but in a way that’s entirely inconsistent with KeeWeb design.

My reason for using KeeWeb instead of KeePass is precisely its design principles. I rarely, if ever, expect to use the “web” part of KeeWeb. The user interface that Antelle has developed is attractive, comfortable and reasonably straightforward to explain to non-geeks.

**Comment by thecount2a** [comment]
For those who don't need GUI and just want a kdbx:// link to just work, I created a somewhat hacky plugin that automates clicking and entering passwords into fields here:

https://github.com/thecount2a/keeweb-linkexternalfile/

This feature should probably be integrated into Keeweb in a better fashion than this. Due to lack of plugin API that serve my purpose, I hook every click and "blur" event within keeweb to see whether there are any links on the screen that need "patched" to properly override their behavior. I understand that this is not good programming practice! If Keeweb adds some infrastructure to define how a plugin could listen for URL click events, this would become much cleaner. Unfortunately, I don't know enough about the structure of Keeweb to propose how this should be done properly. Until then, we have something to start playing with.

**Comment by jeff-cook** [comment]
KeePassXC and Keepass2Android both natively support this. They use a group in the root called AutoOpen. They do not support a GUI to configure. 

Any items in the /AutoOpen group gets loaded. Using the URL and password. 

https://github.com/keepassxreboot/keepassxc/issues/477 
https://github.com/PhilippC/keepass2android/issues/174

**Comment by le-ross** [comment]
Can we support the feature by any means?
